Ansible for installing Galaxy (#1195)
galaxyproject/training-material: topics/admin/tutorials/ansible-galaxy/tutorial.md

contributors:
  - erasche
requirements:
  - type: "internal"
    topic_name: admin
    tutorials:
      - ansible
  -
    title: "A server/VM to deploy Galaxy on"
    type: "none"
  - type: "external"
    title: Ansible setup on your local machine
    link: "https://docs.ansible.com/ansible/latest/installation_guide/intro_installation.html"
  -
    title: Comfort with Ansible roles and tasks
    type: "none"
---

# Overview
{:.no_toc}

Now that you have some familiarity with Ansible and are comfortable running existing playbooks that you've written, we'll move on to installing Galaxy with a playbook.

We want to give you a comprehensive understanding of how the Galaxy installation occurs, but we want to avoid you having to write a "custom" Galaxy installation playbook which you would eventually throw away, in order to use the official playbooks. Given these goals, we will go through the playbook in depth first, and then move to a hands-on portion later. If you are not interested in the inner workings, you can [skip to that section now](#hands-on-installing-galaxy).

> ### Agenda
>
> 1. TOC
> {:toc}
>
{: .agenda}

# Playbook Overview

## Configuration

> The official playbook is extremely configurable, everything that you want to change is exposed as a variable, and then tasks will change behaviour based on that. The [playbook documentation](https://github.com/galaxyproject/ansible-galaxy#role-variables) is the most up-to-date source of documentation for the variables. You should take a minute a read over the variables listed there.

Reviewer: > take a minute a read over the variables

This doesn't sound right.

To proceed from here it is expected that:

- You have [Ansible installed](https://docs.ansible.com/ansible/latest/installation_guide/intro_installation.html) on your local machine
- You have a [hosts file](../ansible/tutorial.html#hosts-file) with the VM or host specified where you will deploy galaxy. We will refer to this group of hosts as "galaxyservers" but you can refer to it with another name if you prefer.

## Requirements

We have codified all of the dependencies you will need into a yaml file that `ansible-galaxy` can install

> ### {% icon hands_on %} Hands-on: Minimal Galaxy Playbook
>
> 1. Create a new file in your working directory called `requirements.yaml` and include the following contents:
>
>    ```yaml
>    - src: galaxyproject.galaxy
>    - src: galaxyproject.postgresql
>    - src: galaxyproject.proftpd
>    - src: https://github.com/usegalaxy-eu/ansible-repos
>      name: galaxyproject.repos
>    - src: geerlingguy.nginx
>    - src: natefoo.postgresql_objects
>    - src: https://github.com/usegalaxy-eu/ansible-role-supervisor

> >      name: geerlingguy.supervisor

Reviewer 2: `usegalaxy-eu.supervisor`?

Author: figured I'd move it eventually?

---
layout: tutorial_hands_on

title: "Galaxy Installation with Ansible"
questions:
- How does the Galaxy Ansible module work internally?
- How can I install a Galaxy server with Ansible
objectives:
- Have an understanding of how Galaxy's Ansible roles are structured and interact with one another
- Be able to use an Ansible playbook to install different flavors of Galaxy for different purposes
time_estimation: "2h"
key_points:
- Basic deployment with Ansible is surprisingly easy
- Complexity can grow over time as your organisation does, no need to start with playbooks like usegalaxy.org
contributors:
  - erasche
requirements:
  - type: "internal"
    topic_name: admin
    tutorials:
      - ansible
  -
    title: "A server/VM to deploy Galaxy on"
    type: "none"
  - type: "external"
    title: Ansible setup on your local machine
    link: "https://docs.ansible.com/ansible/latest/installation_guide/intro_installation.html"
  -
    title: Comfort with Ansible roles and tasks
    type: "none"
---

# Overview
{:.no_toc}

> Now that you have some familiarity with Ansible and are comfortable running existing playbooks that you've written, we'll move on to installing Galaxy with a playbook.

Reviewer: This intro is a bit too tailored for the Admin course, can you rephrase it so it applies also for people reading this on their own (and which may have learnt Ansible somewhere else)?

Author: This tutorial assumes the knowledge contained in the requirements, namely the existing ansible tutorial in the training material.

Reviewer: I've seen that, what I don't like of the present formulation is the "Now that you have..." and "we'll move on" parts, which seem to assume you have just finished the Ansible tutorial, while the reader may have learnt it somewhere else or a long time ago.

What about the following:

```suggestion
This tutorial assumes you have some familiarity with Ansible and are comfortable with writing and running playbooks. Here we'll see how to install a Galaxy server using an Ansible playbook.
```

Author: Thanks for the suggestion, yeah, that fits for me :) Definitely needed to rework that formulation, don't want to put too much pressure on the reader, just that "there is an assumption/belief that you're comfortable with this already."

We want to give you a comprehensive understanding of how the Galaxy installation occurs, but we want to avoid you having to write a "custom" Galaxy installation playbook which you would eventually throw away, in order to use the official playbooks. Given these goals, we will go through the playbook in depth first, and then move to a hands-on portion later. If you are not interested in the inner workings, you can [skip to that section now](#hands-on-installing-galaxy).

> ### Agenda
>
> 1. TOC
> {:toc}
>
{: .agenda}

# Playbook Overview

## Configuration

The official playbook is extremely configurable, everything that you want to change is exposed as a variable, and then tasks will change behaviour based on that. The [playbook documentation](https://github.com/galaxyproject/ansible-galaxy#role-variables) is the most up-to-date source of documentation for the variables. You should take a minute a read over the variables listed there.

> The important variables for our purposes today will be:

Reviewer: ```suggestion
The important variables for this tutorial are:
```

- `galaxy_server_dir`
- `galaxy_commit_id`
- `galaxy_config`

> These are largely self explanatory, a server directory, which commit should be installed, and the Galaxy configuration. We will not Galaxy configuration variables in detail as they are covered sufficiently in the `galaxy.yml` file or the [online documentation](https://docs.galaxyproject.org/en/master/admin/config.html#configuration-options).

Reviewer: ```suggestion
These are largely self explanatory: a server directory, which commit should be installed, and the Galaxy configuration. We will not explain Galaxy configuration variables in detail as they are covered sufficiently in the `galaxy.yml` sample file or the [online documentation](https://docs.galaxyproject.org/en/master/admin/config.html#configuration-options).
```

- [Managing Mutable Setup](#mutable-setup)
- [Managing the Database](#managing-the-database)

### Cloning Galaxy

The [clone](https://github.com/galaxyproject/ansible-galaxy/blob/master/tasks/clone.yml) task is the one which is primarily interesting to us, it downloads Galaxy, using git, to a specific commit.

1. It starts by checking if the Galaxy directory exists, this is done so the tasks will attempt to check the current commit, will not fail in case the folder does not exist.
2. The current commit of the repository on disk is retrieved, in order to decide whether or not Galaxy needs to be updated. This is then reported in a debugging message to the user.
3. Galaxy is updated to the correct commit. If the `galaxy_server_dir` is not yet populated, e.g. on first run, Galaxy will be cloned to the specified location. Otherwise the existing repo will be updated to that commit. The `galaxy_force_checkout` variable allows you to control whether or not any locally made changes will be discarded. This will not affect mutable configuration files as those are not tracked in git.
4. The virtualenv is set up:
	1. An empty virtualenv is created.
	2. Pip is updated within the virtualenv.

> 5. Any `.pyc` files are removed, as this can occasionally result in python loading the cached code, instead of any updates that were made by changing commit. For safety, all of these are removed.

Reviewer: ```suggestion
5. Any `.pyc` files are removed, as this can occasionally result in Python loading the cached code, even if the corresponding `.py` file is no more present at the checked-out commit. For safety, all of these are removed.
```

> With that Galaxy is cloned to disk and is ready to be configured by the next task

Reviewer: ```suggestion
With that Galaxy is cloned to disk and is ready to be configured by the next task.
```
